plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceKey.java

```diff
@@ -10,57 +10,48 @@
  *******************************************************************************/
 package org.eclipse.che.plugin.docker.machine;
 
-import org.eclipse.che.api.machine.server.spi.InstanceKey;
-import org.eclipse.che.dto.server.DtoFactory;
-import org.eclipse.che.dto.shared.JsonStringMap;
+import com.google.common.collect.ImmutableMap;
 
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
+import org.eclipse.che.api.machine.server.impl.InstanceKeyImpl;
+import org.eclipse.che.api.machine.server.spi.InstanceKey;
 
 /**
- * Set of keys that identifies docker image properties
+ * Set of helper method that identifies docker image properties
  *
- * @author andrew00x
+ * @author Sergii Kabashnyuk
  */
-public class DockerInstanceKey implements InstanceKey {
+public class DockerInstanceKey extends InstanceKeyImpl {
     public static final String REPOSITORY = "repository";
     public static final String TAG        = "tag";
     public static final String ID         = "id";
     public static final String REGISTRY   = "registry";
 
-    private final Map<String, String> fields;
-
-    public DockerInstanceKey(InstanceKey instanceKey) {
-        fields = new HashMap<>(4);
-        fields.put(REPOSITORY, instanceKey.getFields().get(REPOSITORY));
-        fields.put(TAG, instanceKey.getFields().get(TAG));
-        fields.put(ID, instanceKey.getFields().get(ID));
-        fields.put(REGISTRY, instanceKey.getFields().get(REGISTRY));
+    public DockerInstanceKey(InstanceKey key) {
+        super(key);
     }
 
     public DockerInstanceKey(String repository, String tag, String id, String registry) {
-        fields = new HashMap<>(4);
-        fields.put(REPOSITORY, repository);
-        fields.put(TAG, tag);
-        fields.put(ID, id);
-        fields.put(REGISTRY, registry);
+        super(ImmutableMap.of(REPOSITORY, repository, TAG, tag, ID, id, REGISTRY, registry));
+    }
+
+    public DockerInstanceKey(String repository, String id, String registry) {
+        super(ImmutableMap.of(REPOSITORY, repository, ID, id, REGISTRY, registry));
     }
 
     public String getRepository() {
-        return fields.get(REPOSITORY);
+        return getFields().get(REPOSITORY);
```

Comment:
You create instance of immutable map on each call of this method. See my comment here https://github.com/codenvy/che-core/pull/575/files#r48469281